### PR TITLE
[PROCS-3651] Create a process agent component

### PR DESCRIPTION
**Pull request by wiyu**
### What does this PR do?
This introduces a new process-agent component which can then be used in either the core agent (linux only) or the process-agent. This will allow for a reduced agent footprint.

The following config `process_config.run_in_core_agent.enabled` has been added to enable this component in the core agent or process-agent. 

The process checks will only allowed to run the core agent if the config is set and the connections(NPM) check is not enabled. The connections check currently requires the process check. This is only a temporary constraint and will be addressed in future work to remove this requirement to allow the process checks to always run in the core agent.

#### Linux
| Configuration                                                                                           | Core Agent         | Process Agent      |
|---------------------------------------------------------------------------------------------------------|--------------------|--------------------|
| run_in_core_agent :white_check_mark:<br>process, container, or process-discovery collection :white_check_mark:                           | :white_check_mark: | :x:                |
| run_in_core_agent :x:<br>process collection :white_check_mark:                                          | :x:                | :white_check_mark: |
| run_in_core_agent :white_check_mark:<br>process collection :white_check_mark:<br>NPM :white_check_mark: | :x:                | :white_check_mark: |
|                                                                                                         |                    |                    |

#### Non-Linux (windows, mac)
The process-agent will continue to run as a stand-alone agent.

### Motivation

This creates a new process-agent component. This component will allow us to reuse the process checks in the core agent and continue using it in the process agent in a reusable component.

https://datadoghq.atlassian.net/browse/PROCS-3651

### Additional Notes

The following two configuration values were considered: `run_in_core_agent` and `run_in_external_agent`. The former was selected as the process checks currently run in an external agent. As we plan on migrating these checks from an external agent to the core agent going from an of to on eases the operators cognitive load. Also as a general practice feature flags should be toggled from false to true.

### Describe how to test/QA your changes
This change is tested via unit tests. This change is currently dormant and not integrated into the agent, so it does not require any manual QA

### Reviewer's Checklist

- [x] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [x] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [x] Changed code has automated tests for its functionality.
- [x] Adequate QA/testing plan information is provided. Except if the `qa/skip-qa` label, with required either `qa/done` or `qa/no-code-change` labels, are applied.
- [x] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: comp/process/agent/agentimpl/agent.go [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+	"github.com/DataDog/datadog-agent/comp/process/agent"
+	"github.com/DataDog/datadog-agent/comp/process/runner"
+	"github.com/DataDog/datadog-agent/comp/process/types"
+	"github.com/DataDog/datadog-agent/pkg/process/checks"
+	"github.com/DataDog/datadog-agent/pkg/util/fxutil"
+	"github.com/DataDog/datadog-agent/pkg/util/optional"
+)
+
+const (
+	agent6DisabledMessage = `process-agent not enabled.
```
Comment (kkhor-datadog) [inline comp/process/agent/agentimpl/agent.go:26]: is there a reason this is specifically called `agent6`?

**Review by kkhor-datadog — COMMENTED** [review]
This can be done in a follow-up, but since this is a pretty significant config, seems worth exposing it in the process agent status output

File: comp/process/agent/agentimpl/agent.go (hunk 1 above)
Comment (wiyu) [reply comp/process/agent/agentimpl/agent.go:26]: This was ported from https://github.com/DataDog/datadog-agent/blob/a958bf57751fb97a8c4777c957053f040f798a22/cmd/process-agent/command/main_common.go#L48-L56

I can rename the const as it's version agnostic.

**Comment by wiyu** [comment]
> This can be done in a follow-up, but since this is a pretty significant config, seems worth exposing it in the process agent status output

@kkhor-datadog - Thank, I'll add this as follow-up for the integration work.

File: comp/process/agent/agentimpl/agent_linux_test.go [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+			expected:             false,
+		},
+		{
+			name:                 "core agent with connections enabled and run in core-agent mode enabled",
+			agentFlavor:          flavor.DefaultAgent,
+			checksEnabled:        true,
+			checkName:            checks.ConnectionsCheckName,
+			runInCoreAgentConfig: true,
+			expected:             false,
+		},
```
[the remark this answers is not in the document]
Comment (wiyu) [reply comp/process/agent/agentimpl/agent_linux_test.go:93]: Unfortunately the mock object is currently is only a single check, but I could look into modifying to be a slice of checks to better model the behavior we expect.

File: comp/process/agent/agentimpl/agent.go [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+	Runner runner.Component
+}
+
+func newProcessAgent(p processAgentParams) optional.Option[agent.Component] {
+	if !agentEnabled(p) {
+		return optional.NewNoneOption[agent.Component]()
+	}
+
+	// Look to see if any checks are enabled, if not, return since the agent doesn't need to be enabled.
+	if !checksEnabled(p.Checks) {
```
Comment (robertjli) [inline comp/process/agent/agentimpl/agent.go:63]: Nit, optional: could this be combined with the populating of `enabledChecks` below?

File: comp/process/agent/agentimpl/agent_linux.go [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+// It will run as a standalone Process-agent if 'run_in_core_agent' is not enabled or a the connections/NPM check is
+// enabled.
+// If 'run_in_core_agent' flag is enabled and the connections/NPM check is not enabled, the process-agent will run in
+// the core agent.
+func agentEnabled(p processAgentParams) bool {
+	runInCoreAgent := p.Config.GetBool("process_config.run_in_core_agent.enabled")
+
+	switch flavor.GetFlavor() {
+	case flavor.ProcessAgent:
+		return !runInCoreAgent
```
Comment (robertjli) [inline comp/process/agent/agentimpl/agent_linux.go:27]: Does this need to return true if the connections check is enabled and runInCoreAgent is true?

File: comp/process/agent/agentimpl/agent_stub.go
Comment (robertjli) [inline comp/process/agent/agentimpl/agent_stub.go:1]: Nit, optional: We've typically named the fallback file `*_fallback.go`, might make it clearer that this isn't included on linux

File: comp/process/agent/component.go [hunk 5]
```diff
@@ -9,7 +9,5 @@ package agent
 // team: processes
 
 // Component represents the no-op Component interface.
```
Comment (ogaca-dd) [inline comp/process/agent/component.go:11]: This comment also need to be updated as the `Component` is not a `no-op`, isn't it?

**Review by ogaca-dd — APPROVED** [review]
LGTM

File: comp/process/agent/agentimpl/agent_linux.go (hunk 4 above)
Comment (wiyu) [reply comp/process/agent/agentimpl/agent_linux.go:42]: Thanks for catching! I've added a test to cover it.

File: comp/process/agent/agentimpl/agent_linux.go [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+	case flavor.DefaultAgent:
+		for _, check := range p.Checks {
+			// the connections check is not supported in the core agent
+			if check.Object().Name() == checks.ConnectionsCheckName && check.Object().IsEnabled() {
+				return false
```
[the remark this answers is not in the document]
Comment (wiyu) [reply comp/process/agent/agentimpl/agent_linux.go]: Added the logging. let me know what you think

**Review by robertjli — APPROVED** [review]
LGTM! 🙇

**Review by xornivore — COMMENTED** [review]
+1 on @robertjli ask to update the description to express complete details and constraints for this change.

File: comp/process/agent/agentimpl/agent_linux.go [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+)
+
+// agentEnabled determines whether the process agent is enabled based on the configuration.
+// The process-agent component on linux can be run in the core agent or as a standalone process-agent
+// depending on the configuration.
+// It will run as a standalone Process-agent if 'run_in_core_agent' is not enabled or a the connections/NPM check is
+// enabled.
+// If 'run_in_core_agent' flag is enabled and the connections/NPM check is not enabled, the process-agent will run in
+// the core agent.
+func agentEnabled(p processAgentParams) bool {
```
[the remark this answers is not in the document]
Comment (wiyu) [reply comp/process/agent/agentimpl/agent_linux.go:22]: So `enabledInCoreAgent` is not quite accurate. The flag shows us if the user has it enabled, but this function is seeing if the process-agent component should be enabled given the configuration and context (agent flavor).

File: comp/process/agent/component.go [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2024-present Datadog, Inc.
+
+// Package agent contains a process-agent component
+package agent
+
+// team: processes
+
+// Component is the process agent component type
+type Component interface {
```
Comment (daniel-taf) [inline comp/process/agent/component.go:12]: Why did this empty interface need to be created? I assume this is an Fx specific pattern.

File: comp/process/agent/agentimpl/agent.go [hunk 9]
```diff
@@ -0,0 +1,106 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2024-present Datadog, Inc.
+
+// Package agentimpl implements a component for the process agent.
```
Comment (daniel-taf) [inline comp/process/agent/agentimpl/agent.go:6]: Are there any other agent components that have been created in the past and are used by the core agent? I'm a bit confused on the mechanism for using this , so some similar code might be helpful.

Comment (wiyu) [reply comp/process/agent/agentimpl/agent.go:6]: So this is based off a POC I did earlier of running this in the core agent. You can see it here and how it's used: https://github.com/DataDog/datadog-agent/pull/20818

File: comp/process/agent/component.go (hunk 8 above)
Comment (wiyu) [reply comp/process/agent/component.go:12]: This isn't specific to fx, but a typical dependency injection pattern. This interface allows us to create other implementations which allows either the application or tests to be wired with different implementations.

You should have a look through the guide here for additional details: https://uber-go.github.io/fx/get-started/

**Review by kkhor-datadog — APPROVED** [review]

File: pkg/config/setup/process_test.go [hunk 10]
```diff
@@ -59,6 +60,10 @@ func TestProcessDefaultConfig(t *testing.T) {
 			key:          "process_config.container_collection.enabled",
 			defaultValue: true,
 		},
+		{
+			key:          "process_config.run_in_core_agent.enabled",
```
[the remark this answers is not in the document]
Comment (wiyu) [reply pkg/config/setup/process_test.go:64]: Added under additional notes. Feel free to debate it! 😆

**Review by xornivore — APPROVED** [review]
:shipit: But maybe `agentEnabled` naming can be improved to make the meaning clearer.

**Pull request merged**